### PR TITLE
[SCHEMA] Add patterns for format validation

### LINKED ISSUE #832: Define valid value formats in schema
I'm thinking it might also be a good idea to create a new `formats.yaml` that defines the different formats used in the schema. Something like:

```yaml
label:
    description: |
        Freeform labels without special characters.
    pattern: ^[0-9a-zA-Z]+$
index:
    description: |
        Non-negative, non-zero integers, optionally prefixed with leading zeros for sortability.
    pattern: ^[0-9]+$
unit:
    description: |
        A unit.
    pattern: ^[0-9a-zA-Z]+$
```

Unfortunately, I don't actually know how most of those patterns would look. Units, for example, can be freeform text if necessary, but optimally they should be valid combinations of the values in the Units appendix. Plus we have relative paths of multiple flavors, along with URIs. I think `uri`, at least, is offloaded to JSON Schema in the validator.

_Originally posted by @tsalo in https://github.com/bids-standard/bids-specification/issues/792#issuecomment-831462364_

EDIT: Labeling the different types of paths, and thus defining patterns for them, will have to wait until after #820 is resolved, at least.

### PR DESCRIPTION
**Pull request by tsalo**
Closes #832.

This currently doesn't include any code that would _use_ the patterns, except for tests.

Changes proposed:
- Add `formats.yaml` file to schema, with regular expressions with which to validate different entity values, metadata types, and metadata string formats. Individual metadata fields should still be allowed to supply a specific "pattern" keyword.

To do:
- [x] write tests (post-#892)
- [x] add rrid pattern (for #827 and the existing `SoftwareRRID`) 
- [ ] review the URI and path patterns

### CODE REVIEW

File: src/schema/formats.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+index:
+  description: |
+    Non-negative, non-zero integers, optionally prefixed with leading zeros for sortability.
+  pattern: ^[0-9]+$
+# Metadata types
+string:
+  description: |
+    A basic string type (not a specific format).
+    This should allow any free-form string *except* "n/a".
+  pattern: ^(?!(n/a)$).*$
```
Comment (tsalo) [inline src/schema/formats.yaml:17]: The "n/a" exception is included for #876.

File: src/schema/formats.yaml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    The validation for this format is minimal.
+  pattern: ^[0-9a-zA-Z/_-\.]+$
+uri:
+  description: |
+    A uniform resource indicator.
+  pattern: ^[a-zA-Z]+:[0-9a-zA-Z/_-\.]+$
+bids_uri:
+  description: |
+    A BIDS uniform resource indicator.
+  pattern: ^bids:[0-9a-zA-Z/_-\.]+$
```
Comment (tsalo) [inline src/schema/formats.yaml:66]: The patterns for all four of the path/URI formats should be revised. I don't know enough about the rules to define solid patterns just yet.

File: src/schema/formats.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+number:
+  description: |
+    A number.
+  pattern: ^[+-]?([0-9]+([.][0-9]*)?|[.][0-9]+)$
+boolean:
+  description: |
+    A boolean.
+  pattern: ^(true|false)$
+# String patterns
+date:
```
Comment (tsalo) [inline src/schema/formats.yaml:31]: `date` is included in the schema as a "pattern" rather than a "format". That was just an oversight on my part. I would fix it here, but that would cause a conflict with #883, so I will address it in that PR.

File: src/schema/formats.yaml (hunk 2 above)
Comment (tsalo) [reply src/schema/formats.yaml:72]: The same goes for `stimuli_relative` (added after the initial comment).

File: src/schema/formats.yaml (hunk 3 above)
Comment (tsalo) [reply src/schema/formats.yaml:31]: Fixed in e9c4f5f (part of #883).

**Review by Remi-Gau — COMMENTED** [review]
This will for sure be superuseful. I don't speak regex fluently so not sure I am the best to review this though.

**Review by rwblair — APPROVED** [review]
Structure makes sense to me. Eyeballing the regex it seems ok, but the best way to verify it is for it to be used by the validator in tests.

**Comment by tsalo** [comment]
Thanks for taking a look. I'm thinking that I should try writing some unit tests for the schema code, which will necessitate reorganizing it into a proper package. I will open a PR to do that after #883 is merged, so that there aren't too many merge conflicts.

File: src/schema/formats.yaml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  description: |
+    Non-negative, non-zero integers, optionally prefixed with leading zeros for sortability.
+    An index may not be all zeros.
+  pattern: ^[0-9]*[1-9]+[0-9]*$
+# Metadata types
+string:
+  description: |
+    The basic string type (not a specific format).
+    This should allow any free-form string *except* "n/a".
+  pattern: ^(?!(n/a)$).*$
```
Comment (sappelhoff) [inline src/schema/formats.yaml:18]: This is related to #876 

> This should allow any free-form string *except* "n/a".

I am not sure whether disallowing `n/a` as a string for free-form text fields would break some datasets. For example for the required `EEGReference` field, users could currently pass `n/a` according to bids-validator:

https://github.com/bids-standard/bids-validator/blob/75c54e4f7a17bf58c7cfc47c9698eb4e0841315b/bids-validator/validators/json/schemas/eeg.json#L25

Whether or not that's a good idea should be discussed in #876, this is just a note that what you introduce here has not been discussed and has not been formally part of the spec :thinking:

**Comment by yarikoptic** [comment]
I guess the next step for this pr is adding some tests as @tsalo [was planning todo](https://github.com/bids-standard/bids-specification/pull/885#issuecomment-934535138) (and #833 is merged) or that would be postponed till later?

**Comment by tsalo** [comment]
I'd prefer to add the tests before merging. When they have time, Anthony and Chris are planning to take a look at #892. Once that's merged, I can add tests for the formats to this PR.

File: src/schema/objects/formats.yaml [hunk 5]
```diff
@@ -0,0 +1,99 @@
+---
+# This file defines valid patterns for different formats
+# Entity patterns
+label:
+  description: |
+    Freeform labels without special characters.
+  pattern: ^[0-9a-zA-Z]+$
```
Comment (yarikoptic) [inline src/schema/objects/formats.yaml:7]: what is the motivation for explicit anchoring?
We with @TheChymera we are thinking to establish regex based validation of BIDS filenames for DANDI project... for that, we likely just to compose a list of regexes for a valid BIDS filename.  Then, `<label>` would be replaced with regex from here.  If it is anchored -- it would require us to strip those first.  I think anchors should be explicitly added in the code whenever processing already "extracted" label.

Comment (tsalo) [reply src/schema/objects/formats.yaml:7]: I believe I kept the anchors from the validator expressions. I can definitely remove them here, and then the validator would just add them back I think.

Comment (tsalo) [reply src/schema/objects/formats.yaml:7]: Are quotes a problem? Some of the regular expressions seem to cause problems with YAML if I don't wrap them in quotes.

File: src/schema/objects/formats.yaml [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+integer:
+  name: Integer
+  description: |
+    An integer which may be positive or negative.
+  pattern: '[+-]?\d+'
+number:
+  name: Number
+  description: |
+    A number which may be an integer or float, positive or negative.
+  pattern: '[+-]?([0-9]+([.][0-9]*)?|[.][0-9]+)'
```
Comment (effigies) [inline src/schema/objects/formats.yaml:31]: I would expect scientific notation (e.g., `1e3`, `-2.1E+5`) to be permissible.

```suggestion
  pattern: '[+-]?([0-9]+([.][0-9]*)?|[.][0-9]+)([eE][+-]?[0-9]+)?'
```

File: src/schema/objects/formats.yaml [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  description: |
+    A datetime in the form `"YYYY-MM-DDThh:mm:ss[.000000][Z]"`,
+    where [.000000] is an optional subsecond resolution between 1 and 6 decimal points,
+    and [Z] is an optional, valid timezone code.
+  pattern: '[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}?(\.[0-9]{1,6})?([A-Z]{3,4})'
+time:
+  name: Time
+  description: |
+    A time in the form `"hh:mm:ss"`.
+  pattern: '(?:2[0-3]|[01][0-9]):[0-5][0-9]:[0-5][0-9]'
```
Comment (effigies) [inline src/schema/objects/formats.yaml:56]: Should we permit `1:00:00`? Or just `01:00:00`?

```suggestion
  pattern: '(?:2[0-3]|[01]?[0-9]):[0-5][0-9]:[0-5][0-9]'
```

**Review by effigies — COMMENTED** [review]
Couple quick thoughts.

File: src/schema/objects/formats.yaml (hunk 6 above)
Comment (tsalo) [reply src/schema/objects/formats.yaml:31]: ohhhh good catch!

File: src/schema/objects/formats.yaml (hunk 7 above)
Comment (tsalo) [reply src/schema/objects/formats.yaml:56]: I think that's reasonable. Thanks!

File: src/schema/objects/formats.yaml [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  name: Research resource identifier
+  description: |
+    A [research resource identifier](https://scicrunch.org/resources).
+  pattern: 'RRID:.+_.+'
+uri:
+  name: Uniform resource indicator
+  description: |
+    A uniform resource indicator.
+  pattern: '(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?'
+bids_uri:
```
Comment (yarikoptic) [inline src/schema/objects/formats.yaml:107]: might be worth to sort the entries to ease locating any particular one

File: tools/schemacode/schemacode/tests/test_schema.py [hunk 9]
```diff
@@ -27,3 +27,101 @@ def test_object_definitions(schema_obj):
...
+            "05/01/2022",  # MM/DD/YYYY or DD/MM/YYYY
+        ],
+        "datetime": [
+            "2022-01-05T13:16:30.1000005",  # too many decimal points
+            "2022-01-05T13:16:30U",  # time zone too short
+            "2022-01-05T13:16:30UTCUTC",  # time zone too long
+            "2022-01-05T34:10:10",  # invalid time
+        ],
+        "time": [
+            "34:10:10",  # invalid time
```
Comment (yarikoptic) [inline tools/schemacode/schemacode/tests/test_schema.py:104]: I would have added some extra boundary conditions
```suggestion
            "34:10:10",  # invalid time
            "24:00:00",  # should be 00:00:00
            "00:60:00",  # should be 01:00:00
            "00:00:60",  # should be 00:01:00
            "01:23",     # lacks either hours or seconds
```

File: tools/schemacode/schemacode/tests/test_schema.py [hunk 10]
```diff
@@ -27,3 +27,101 @@ def test_object_definitions(schema_obj):
...
+            "stimuli/path/file.txt",
+        ],
+        "dataset_relative": [
+            "/path/with/starting/slash/file.txt",
+        ],
+        "participant_relative": [
+            "/path/with/starting/slash/file.txt",
+            "sub-01/path/file.txt",
+        ],
+        "rrid": [],
```
Comment (yarikoptic) [inline tools/schemacode/schemacode/tests/test_schema.py:118]:
```suggestion
        "rrid": [
            "RRID:",  # empty one
        ],
```

File: tools/schemacode/schemacode/tests/test_schema.py [hunk 11]
```diff
@@ -27,3 +27,101 @@ def test_object_definitions(schema_obj):
...
+        "uri": [],
```
Comment (yarikoptic) [inline tools/schemacode/schemacode/tests/test_schema.py:119]:
```suggestion
        "uri": [
            "ftp://"  # lacks anything but protocol
        ],
```
didn't check if that would be caught by regex, but clearly such URI would be quite useless

File: tools/schemacode/schemacode/tests/test_schema.py [hunk 12]
```diff
@@ -27,3 +27,101 @@ def test_object_definitions(schema_obj):
...
+        "bids_uri": [],
+    }
+    for pattern, test_list in BAD_PATTERNS.items():
+        pattern_format = schema_obj["objects"]["formats"][pattern]["pattern"]
+        search_pattern = "^" + pattern_format + "$"
```
Comment (yarikoptic) [inline tools/schemacode/schemacode/tests/test_schema.py:124]: I guess I just love f-strings too much?

```suggestion
        search_pattern = f"^{pattern_format}$"
```

File: tools/schemacode/schemacode/tests/test_schema.py [hunk 13]
```diff
@@ -27,3 +27,101 @@ def test_object_definitions(schema_obj):
...
+        search = re.compile(search_pattern)
+        for test_string in test_list:
+            assert bool(
+                search.fullmatch(test_string)
+            ), f"'{test_string}' is not a valid match for the pattern '{search.pattern}'"
+
+    # Check that invalid strings do not match the search pattern.
+    BAD_PATTERNS = {
+        "label": ["test_01", "!", "010101-"],
```
Comment (yarikoptic) [inline tools/schemacode/schemacode/tests/test_schema.py:87]:
```suggestion
        "label": ["test_01", "!", "010101-", "01-01", "-01"],
```
just to have one with `-` in the middle or at the beginning

File: tools/schemacode/schemacode/tests/test_schema.py [hunk 14]
```diff
@@ -27,3 +27,101 @@ def test_object_definitions(schema_obj):
...
+        "index": ["test", "0.1", "0-1", "0_1"],
+        "string": [],
+        "integer": ["3.14", "-3.14", "1.", "-1.", "string", "s1", "1%"],
```
Comment (yarikoptic) [inline tools/schemacode/schemacode/tests/test_schema.py:90]:
```suggestion
        "integer": ["3.14", "-3.14", "1.", "-1.", "string", "s1", "1%", "one"],
```
AI is not supported!

**Review by yarikoptic — COMMENTED** [review]
I just love testing boundary conditions I guess ;)

File: src/schema/objects/formats.yaml (hunk 8 above)
Comment (tsalo) [reply src/schema/objects/formats.yaml:107]: Good idea. I'll sort them after merging in your recommendations.

File: tools/schemacode/schemacode/tests/test_schema.py [hunk 15]
```diff
@@ -27,3 +27,109 @@ def test_object_definitions(schema_obj):
...
+            "2022-01-05T13:16:30.000005",  # up to 6 decimal points
+            "2022-01-05T13:16:30UTC",  # timezones are allowed
+            "2022-01-05T13:16:30.05UTC",
+        ],
+        "time": [
+            "13:16:30",
+            "09:00:00",
+            "9:00:00",  # leading zeros are not required for hours
+        ],
+        "unit": ["any string is valid."],
```
Comment (yarikoptic) [inline tools/schemacode/schemacode/tests/test_schema.py:64]: oh, even with spaces and unicodes?  I thought we had discussion and there was a decision on some unicodes to be ok (but not any), then with #411 CMIXF-12 ascii recommended. https://github.com/bids-standard/bids-specification/blob/HEAD/src/99-appendices/05-units.md mentions `µ` (could be only as prefix) `Ω` (only as suffix) and `°` (all by itself?).  So may be regex and test here could be made more stringent?  well, aforementioned page gives example just as "NOT RECOMMENDED" instead of "WRONG": `volts per microsecond` so may be my whining is all in vein?  But then I would restrict to composition of alphanumerics, space, `-` and those 3 unicodes, and thus disallow everything else.

**Comment by tsalo** [comment]
It looks like `ftp://` passes validation, unfortunately. Not sure how to update the regex appropriately.

File: tools/schemacode/schemacode/tests/test_schema.py (hunk 15 above)
Comment (tsalo) [reply tools/schemacode/schemacode/tests/test_schema.py:64]: TBH I allowed any string because I don't know much about the unit validation. Do you mind if we allow any string in this PR and make sure to follow up with a more accurate regex before the next release? Assuming we merge this after 1.7.0 is released, so there'll be a bit of time between 1.7.0 and 1.8.0.

Comment (yarikoptic) [reply tools/schemacode/schemacode/tests/test_schema.py:64]: I would not mind. Please just create an issue for that so we do not forget?

**Comment by yarikoptic** [comment]
> It looks like `ftp://` passes validation for the `uri` format, unfortunately. Not sure how to update the regex appropriately.

I can try later to create one following the "Syntax" of https://en.wikipedia.org/wiki/Uniform_Resource_Identifier which would demand to have a non-degenerate `authority` after //

File: tools/schemacode/schemacode/tests/test_schema.py (hunk 15 above)
Comment (tsalo) [reply tools/schemacode/schemacode/tests/test_schema.py:64]: Done. #999.

**Review by sappelhoff — APPROVED** [review]

**Comment by tsalo** [comment]
Now that I have two approvals, I'm going to merge. I think we can keep improving these patterns (i.e., #999) before the next release.
